app/handler/api.go

```diff
@@ -1,204 +1,73 @@
-	if err != nil {
-		handleErr(ctx, 500, err)
-		return
-	}
-	post := &model.Post{Id: int64(id)}
-	err = post.GetPostById()
-	if err != nil {
-		handleErr(ctx, 404, err)
-		return
-	}
-	ctx.JSONIndent(post, "", "  ")
+func NewErrorStatusJSON(msgs ...string) APIStatusJSON {
+	msg := strings.Join(msgs, " ")
+	return APIStatusJSON{Status: "error", Message: msg}
 }
 
-// APIPostSlugHandler retrieves the post with the given slug.
-func APIPostSlugHandler(ctx *golf.Context) {
-	slug := ctx.Param("slug")
-	post := new(model.Post)
-	err := post.GetPostBySlug(slug)
-	if err != nil {
-		handleErr(ctx, 404, err)
-		return
-	}
-	ctx.JSONIndent(post, "", "  ")
+func NewSuccessStatusJSON(msgs ...string) APIStatusJSON {
+	msg := strings.Join(msgs, " ")
+	return APIStatusJSON{Status: "success", Message: msg}
 }
 
-// APITagsHandler retrieves all the tags.
-func APITagsHandler(ctx *golf.Context) {
-	tags := new(model.Tags)
-	err := tags.GetAllTags()
-	if err != nil {
-		handleErr(ctx, 404, err)
-		return
-	}
-	ctx.JSONIndent(tags, "", "  ")
+func NewAPISuccessResponse(data interface{}, msgs ...string) APIResponseBodyJSON {
+	status := NewSuccessStatusJSON(msgs...)
+	return APIResponseBodyJSON{Data: data, Status: status}
 }
 
-// APITagHandler retrieves the tag with the given id.
-func APITagHandler(ctx *golf.Context) {
-	id, err := strconv.Atoi(ctx.Param("id"))
-	if err != nil {
-		handleErr(ctx, 500, err)
-		return
-	}
-	tag := &model.Tag{Id: int64(id)}
-	err = tag.GetTag()
-	if err != nil {
-		handleErr(ctx, 404, err)
-		return
-	}
-	ctx.JSONIndent(tag, "", "  ")
-}
-
-// APITagSlugHandler retrieves the tag(s) with the given slug.
-func APITagSlugHandler(ctx *golf.Context) {
-	slug := ctx.Param("slug")
-	tags := &model.Tag{Slug: slug}
-	err := tags.GetTagBySlug()
-	if err != nil {
-		handleErr(ctx, 500, err)
-		return
+// APIDocumentationHandler shows which routes match with what functionality,
+// similar to https://api.github.com
+func APIDocumentationHandler(routes map[string]map[string]interface{}) golf.HandlerFunc {
+	return func(ctx *golf.Context) {
+		routes["GET"]["api_documentation_url"] = "/api"
+		ctx.JSONIndent(map[string]interface{}{"request_method": routes}, "", "  ")
 	}
-	ctx.JSONIndent(tags, "", "  ")
 }
 
-// APIUsersHandler retrieves all users.
-func APIUsersHandler(ctx *golf.Context) {
+// handleErr sends the staus code and error message formatted as JSON.
+func handleErr(ctx *golf.Context, statusCode int, err error) {
 	ctx.JSONIndent(map[string]interface{}{
-		"message": "Not implemented",
+		"statusCode": statusCode,
+		"error":      err.Error(),
 	}, "", "  ")
 }
 
-// APIUserHandler retrieves the user with the given id.
-func APIUserHandler(ctx *golf.Context) {
-	id, err := strconv.Atoi(ctx.Param("id"))
+func (status APIStatusJSON) Serialize() []byte {
+	serializedStatus, err := json.Marshal(status)
 	if err != nil {
-		handleErr(ctx, 500, err)
-		return
+		utils.LogOnError(err, "Unable to serialize status.", true)
+		return []byte("")
```

Comment:
Would it also be good to serialize the error into JSON if it errors, like the `handleErr` message did? I realize that the chance of actually hitting this error is really really low, but I think for API consumers it would feel weird to make a request and get an empty response instead of an error. On the other hand, since we only return a `[]byte`, you wouldn't be able to tell whether serialization succeeded or not, since the contents of that byte slice could be the expected JSON or the error JSON, and I don't think it'd be good to have that kind of uncertainty. Hmm... maybe it'd be better to return an `err` along with the serialized error JSON, that way we get the error as JSON, and can check for it when this function is called.

Either way, I'm being super nit-picky with his haha, there's probably no need to make any change from the way it is now.